File: pytests/test_df_fit.py
#!/usr/bin/python

'''
    This example demonstrates how to find best-fit parameters of an action-based
    distribution function that matches the given N-body snapshot.

    The N-body model itself corresponds to a spherically-symmetric isotropic
    Hernquist profile, and we fit it with a double-power-law distribution function
    of Posti et al.2015. We use the exact potential (i.e., do not compute it
    from the N-body model itself, nor try to vary its parameters, although
    both options are possible), and compute actions for all particles only once.
    Then we scan the parameter space of DF, finding the maximum of the likelihood
    function with a multidimensional minimization algorithm.
    This takes a few hundred iterations to converge.

    This Python script is almost equivalent to the C++ test program test_df_fit.cpp,
    up to the difference in implementation of Nelder-Mead minimization algorithm.

    Additionally, we use the MCMC implementation EMCEE to explore the confidence
    intervals of model parameters around their best-fit values
'''
import agama, numpy
from scipy.optimize import minimize

# compute log-likelihood of DF with given params against an array of points
def model_likelihood(params, points):
    print "J0=%6.5g, alpha=%6.5g, ar=%6.5g, az=%6.5g, beta=%6.5g: " \
        % (params['j0'], params['alpha'], params['ar'], params['az'], params['beta']),
    try:
        dpl = agama.DistributionFunction(params)
        norm = dpl.total_mass()
        sumlog = numpy.sum( numpy.log(dpl(points)/norm) )
        print "LogL=%8g, norm=%6.5g" % (sumlog, norm)
        return sumlog
    except ValueError as err:
        print "Exception ", err
        return -1000.*len(points)

# convert from parameter space to DF params: note that we apply
# some non-trivial scaling to make the life easier for the minimizer
def dfparams(args):
    return dict(
        type  = 'DoublePowerLaw',
        alpha = args[0],
        beta  = args[1],
        j0    = numpy.exp(args[2]),  # will always stay positive
        ar    = 3./(2+args[3])*args[3],
        az    = 3./(2+args[3]),
        aphi  = 3./(2+args[3]),  # ensure that sum of ar*Jr+az*Jz+aphi*Jphi doesn't depend on args[3]
        br    = 1.,
        bz    = 1.,
        bphi  = 1.,
        jcore = 0.,
        norm  = 1.)

# function to minimize
def model_search_fnc(args, actions):
    return -model_likelihood(dfparams(args), actions)

# function to maximize
def model_search_emcee(args, actions):
    return model_likelihood(dfparams(args), actions)

def main():
    pot = agama.Potential(type="Dehnen", mass=1, scaleRadius=1.)
    actf= agama.ActionFinder(pot)
    particles = numpy.loadtxt("../temp/hernquist.dat", skiprows=1)
    actions = actf(particles[:,:6])

    # do a parameter search to find best-fit distribution function describing these particles
    initparams = numpy.array([2.0, 4.0, 0.0, 1.0])
    result = minimize(model_search_fnc, initparams, args=(actions,), method='Nelder-Mead', 
        options=dict(maxiter=1000, maxfev=500, disp=True))
    print 'result=',result.x, result.message

    # explore the parameter space around the best-fit values using the MCMC chain
    import emcee, matplotlib.pyplot as plt, triangle
    print 'Starting MCMC'
    ndim = len(initparams)
    nwalkers = 32
    # initial coverage of parameter space
    initwalkers = [initparams + 0.1*numpy.random.randn(ndim) for i in range(nwalkers)]
    sampler = emcee.EnsembleSampler(nwalkers, ndim, model_search_emcee, args=(actions,))
    sampler.run_mcmc(initwalkers, 300)

    # show the time evolution of parameters carried by the ensemble of walkers (time=number of MC steps)
    fig,axes = plt.subplots(ndim, 1, sharex=True)
    for i in range(ndim):
        axes[i].plot(sampler.chain[:,:,i].T, color='k', alpha=0.4)
    fig.tight_layout(h_pad=0.)
    plt.show()

    # show the posterior distribution of parameters
    num_burnin = 100   # number of initial steps to discard
    samples = sampler.chain[:, num_burnin:, :]. reshape((-1, ndim))
    triangle.corner(samples, \
        labels=[r'$\alpha$', r'$\beta$', r'ln($J_0$)', r'$a_r/a_{z,\phi}$'], 
        quantiles=[0.16, 0.5, 0.84], truths=result.x)
    plt.show()
    print "Acceptance fraction: ", numpy.mean(sampler.acceptance_fraction)  # should be in the range 0.2-0.5
    print "Autocorrelation time: ", sampler.acor  # should be considerably shorter than the total number of steps

main()

File: tests/test_df_fit.cpp
/** \file    test_df_fit.cpp
    \author  Eugene Vasiliev
    \date    Aug-Nov 2015

    This example demonstrates how to find best-fit parameters of an action-based
    distribution function that matches the given N-body snapshot.

    The N-body model itself corresponds to an axisymmetric density profile generated
    self-consistently from a double-power-law distribution function (Posti et al.2015),
    with the `test_selfconsistentmodel` program.
    The test is supposed to recover the parameters of DF used to create the N-body model
    (we fit it with the same functional form but unknown a priori parameters).
    We compute the potential from the N-body model itself, and compute actions
    for all particles only once.
    Then we scan the parameter space of DF, finding the maximum of the likelihood
    function with a multidimensional minimization algorithm.
    This takes a few hundred iterations to converge.
*/
#include "potential_sphharm.h"
#include "actions_staeckel.h"
#include "df_halo.h"
#include "particles_io.h"
#include "math_fit.h"
#include "math_core.h"
#include "debug_utils.h"
#include <iostream>
#include <cmath>
#include <stdexcept>

typedef std::vector<actions::Actions> ActionArray;

/// compute log-likelihood of DF with given params against an array of points
double modelLikelihood(const df::DoublePowerLawParam& params, const ActionArray& points)
{
    std::cout << "J0="<<params.j0<< //", Jcore="<<params.jcore<<
        ", alpha="<<params.alpha<<", ar="<<params.ar<<", az="<<params.az<<", aphi="<<params.aphi<<
        ", beta=" <<params.beta <<", br="<<params.br<<", bz="<<params.bz<<", bphi="<<params.bphi<<": ";
    double sumlog = 0;
    try{
        df::DoublePowerLaw dpl(params);
        double norm = dpl.totalMass(1e-4, 100000);
        for(unsigned int i=0; i<points.size(); i++) {
            double val = dpl.value(points[i]);
            sumlog += log(val/norm);
        }
        std::cout << "LogL="<<sumlog<<", norm="<<norm<< std::endl;
        return sumlog;
    }
    catch(std::invalid_argument& e) {
        std::cout << "Exception "<<e.what()<<"\n";
        return -1000.*points.size();
    }
}

/// convert from parameter space to DF params: note that we apply
/// some non-trivial scaling to make the life easier for the minimizer
df::DoublePowerLawParam dfparams(const double vars[])
{
    df::DoublePowerLawParam params;
    params.jcore = 0;
    params.alpha = vars[0];
    params.beta  = vars[1];
    params.j0    = vars[2];
    params.ar    = 3./(1+vars[3]+vars[4])*vars[3];
    params.az    = 3./(1+vars[3]+vars[4])*vars[4];
    params.aphi  = 3./(1+vars[3]+vars[4]);  // ar+az+aphi = 3
    params.br    = 3./(1+vars[5]+vars[6])*vars[5];
    params.bz    = 3./(1+vars[5]+vars[6])*vars[6];
    params.bphi  = 3./(1+vars[5]+vars[6]);
    params.norm  = 1.;
    return params;
}

/// function to be minimized
class ModelSearchFnc: public math::IFunctionNdim{
public:
    ModelSearchFnc(const ActionArray& _points) : points(_points) {};
    virtual void eval(const double vars[], double values[]) const
    {
        values[0] = -modelLikelihood(dfparams(vars), points);
    }
    virtual unsigned int numVars() const { return 7; }
    virtual unsigned int numValues() const { return 1; }
private:
    const ActionArray& points;
};

int main(){
    particles::PointMassArrayCar particles;
    readSnapshot("sampled_model.txt", units::ExternalUnits(), particles);
    potential::PtrPotential pot(new potential::SplineExp(20, 4, particles, coord::ST_AXISYMMETRIC, 1.));
    const actions::ActionFinderAxisymFudge actf(pot);
    ActionArray particleActions(particles.size());
    for(unsigned int i=0; i<particles.size(); i++) {
        particleActions[i] = actf.actions(toPosVelCyl(particles.point(i)));
        if(!math::isFinite(particleActions[i].Jr+particleActions[i].Jz)) {
            std::cout << particleActions[i] <<" for "<< particles.point(i) <<"\n";
            particleActions[i] = particleActions[i>0?i-1:0];  // put something reasonable (unless i==0)
        }
    }

    // do a parameter search to find best-fit distribution function describing these particles
    ModelSearchFnc fnc(particleActions);
    const double initparams[] = {2.0, 4.0, 1.0, 1.0, 1.0, 1.0, 1.0};
    const double stepsizes[]  = {0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1};
    const int maxNumIter = 1000;
    const double toler   = 1e-4;
    double bestparams[7];
    int numIter = math::findMinNdim(fnc, initparams, stepsizes, toler, maxNumIter, bestparams);
    std::cout << numIter << " iterations\n";

    return 0;
}
